repository/client.go

```diff
@@ -281,92 +307,131 @@ func (sc secretClient) UpdateClientData(ctx context.Context, update clientData)
-	watcher, err := client.Watch(ctx, metav1.SingleObject(metav1.ObjectMeta{Name: name}))
+func registerEventHandler(ctx context.Context, informer sharedInformer, kind metav1.Object, name string) (<-chan clientWatchResult, error) {
+	watchCh := make(chan clientWatchResult)
+	_, err := informer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		UpdateFunc: updateHandler(kind, watchCh, name),
+	})
 	if err != nil {
-		return nil, fmt.Errorf("could not watch '%s' in cluster: %w", name, handleError(err))
+		return nil, fmt.Errorf("failed to register event handler for kind %T: %w", kind, err)
 	}
 
-	resultChan := make(chan clientWatchResult)
-
 	go func() {
-		defer close(resultChan)
-		for {
-			select {
-			case <-ctx.Done():
-				watcher.Stop()
-				return
-			case result, open := <-watcher.ResultChan():
-				if !open {
-					return
-				}
-
-				resultChan <- handleWatchEvent(name, result)
-			}
-		}
+		informer.Run(ctx.Done())
 	}()
 
-	return resultChan, nil
+	return watchCh, nil
 }
 
-func handleWatchEvent(cfgName string, event watch.Event) clientWatchResult {
-	if event.Type == watch.Error {
-		return clientWatchResult{
-			dataStr:           "",
-			persistentContext: "",
-			err:               fmt.Errorf("error result in watcher for config '%s'", cfgName),
-		}
-	}
-
-	switch r := event.Object.(type) {
-	case *v1.Secret:
-		dataBytes, ok := r.Data[dataKeyName]
-		if !ok {
-			return clientWatchResult{
-				dataStr:           "",
-				persistentContext: "",
-				err:               fmt.Errorf("could not find data for key %s in secret %s", dataKeyName, cfgName),
+func updateHandler(kind metav1.Object, watchCh chan clientWatchResult, name string) func(prevObj interface{}, newObj interface{}) {
+	return func(prevObj, newObj interface{}) {
+		prevCast, newCast, err := cast(kind, prevObj, newObj)
+		if err != nil {
+			watchCh <- clientWatchResult{
+				err: fmt.Errorf("failed to cast objects to desired kind: %w", err),
 			}
+			return
 		}
 
-		return clientWatchResult{
-			dataStr:           string(dataBytes),
-			persistentContext: r.GetResourceVersion(),
-			err:               nil,
+		if prevCast.GetName() != newCast.GetName() {
+			// this should never happen
+			return
 		}
-	case *v1.ConfigMap:
-		dataString, ok := r.Data[dataKeyName]
-		if !ok {
-			return clientWatchResult{
-				dataStr:           "",
-				persistentContext: "",
-				err:               fmt.Errorf("could not find data for key %s in configmap %s", dataKeyName, cfgName),
-			}
+
+		if prevCast.GetName() != name {
+			// this is not the resource we want to watch, skip
+			return
 		}
 
-		return clientWatchResult{
-			dataStr:           dataString,
-			persistentContext: r.GetResourceVersion(),
-			err:               nil,
+		var newDataString string
+		var prevDataString string
+		switch kind.(type) {
+		case *v1.Secret:
+			prevDataString, newDataString, err = getSecretDataStrings(newCast, prevCast, name)
```

Comment:
I think it is confusing to have arguments in order `new`, `prev` and the return types the other way around. Especially since they have the same type.